### ISSUE TITLE
[FEA] Adding __cuda_array_interface__ to dataframes

### ISSUE THREAD

**Issue by jakirkham** [rapidsai/cudf#2776]
**Is your feature request related to a problem? Please describe.**

I'd like to be able to run the following code.

```python
df = cudf.DataFrame([("fea0", [0, 1]), ("fea1", [1, 2])])
df.__cuda_array_interface__
```

This would make it easier to use cuDF `DataFrame` objects with other libraries that support the `__cuda_array_interface__` protocol.

**Describe the solution you'd like**

It would be nice to have `__cuda_array_interface__` added to `DataFrame` objects.

**Describe alternatives you've considered**

One could use `as_gpu_matrix` to get a Numba device array first (as shown below), which then supports this protocol.

```python
df.as_gpu_matrix().__cuda_array_interface__
```

That said, this means that other library code needs to call `.as_gpu_matrix()` and may need to special case cuDF DataFrames to do so. Bypassing this step allows other library authors to just rely on `__cuda_array_interface__` on GPU objects and use these to streamline their handling.

**Comment by harrism** [comment]
Could you just implement `df.__cuda_array_interface__` as a wrapper that calls `df.as_gpu_matrix().__cuda_array_interface__`?

I would think that just like dlpack, `df.__cuda_array_interface__` necessarily requires converting to a dense matrix first. The reason is that a DF has a separate data pointer per column, whereas `__cuda_array_interface__` has a single data pointer if I'm not mistaken.  So it makes sense to just abstract matrix conversion.

**Comment by jakirkham** [comment]
That could be an option. Would the pointer change between invocations?

Interesting. Thanks for the context.

The [Python buffer protocol]( https://docs.python.org/3/c-api/buffer.html ), which `__cuda_array_interface__` is taking inspiration from, allowed [another layer of indirection via nested pointers]( https://docs.python.org/3/c-api/buffer.html#pil-style-shape-strides-and-suboffsets ). Would having this kind of functionality in `__cuda_array_interface__` help?

**Comment by harrism** [comment]
Not sure. if it allows an array of pointers then I guess so.

**Comment by jakirkham** [comment]
Thoughts on this @kkraus14?

**Comment by beckernick** [comment]
@jakirkham @kkraus14 , perhaps we should consider enhancing `as_gpu_matrix` to (optionally?) identify a common dtype and upcast accordingly. Dataframes with both datetimes and numeric/bool dtypes might require some special casing, but I think this would be generally useful (such as for #2691 )

**Comment by kkraus14** [comment]
I'm not a fan of this. One of the conditions of `__cuda_array_interface__` is to have a contiguous array. Each of the column's data in our DataFrame is a separate contiguous allocation with a columnar layout, which means that we'd need to make a new allocation and reformat the data for it to work with a whole DataFrame. This adds additional constraints where we'd then need a place to "cache" the 2d dense array for it to stay in scope. I think having a `.values` operator that returns a cupy array that then exposes a `__cuda_array_interface__` will be a better solution moving forward, but there's still scoping concerns.

**Comment by harrism** [comment]
@kkraus14 there are several ideas proposed in this thread, so your "this" that you are not a fan of is unclear. Which part don't you like?  

Naive question: can __`cuda_array_interface` invoke a function? Or is it just a data member?

**Comment by kkraus14** [comment]
> @kkraus14 there are several ideas proposed in this thread, so your "this" that you are not a fan of is unclear. Which part don't you like?

I'm not a fan of exposing `__cuda_array_interface__` against DataFrame objects in general. It gives a false impression of the underlying memory / data layout and is confusing from my perspective.

> Naive question: can __cuda_array_interface__ invoke a function? Or is it just a data member?

It can invoke a function

**Comment by harrism** [comment]
I agree with that. I guess it makes sense to provide Series.__cuda_array_interface__ and as_gpu_matrix().__cuda_array_interface__ (which I guess we have). This way there is no false impression of contiguous memory where it is not.

**Comment by jakirkham** [comment]
Maybe it would be useful to organize the questions a bit. First what do you think about adjusting `__cuda_array_interface__` to [handle a degree of indirection]( https://github.com/rapidsai/cudf/issues/2776#issuecomment-530173781 ) (like an array of pointers that must also be dereferenced)? Second would this adjustment make it easier to use cuDF DataFrames with `__cuda_array_interface__`? Third if this is all viable, what are your thoughts on this overall strategy for using `__cuda_array_interface__` with cuDF DataFrames?

**Comment by kkraus14** [comment]
> Maybe it would be useful to organize the questions a bit.
> 
> First what do you think about adjusting `__cuda_array_interface__` to handle a degree of indirection (like an array of pointers that must also be dereferenced)?

I don't like the added complexity at all and wouldn't want to support having to have pointer chasing functions in order to be able to ingest data from it.

> Second would this adjustment make it easier to use cuDF DataFrames with __cuda_array_interface__?

Yes and No. Yes in that it would give us a way to actually represent DataFrames in a single `__cuda_array_interface__` output, no because we'd also need to do this for the null bitmask and then it becomes even more complicated because you're pointer chasing for bits.

> Third if this is all viable, what are your thoughts on this overall strategy for using __cuda_array_interface__ with cuDF DataFrames?

I'm not a fan of exposing __cuda_array_interface__ against DataFrame objects in general. I think there should be a separate protocol for DataFrames, but I haven't given much thought to what it would look like. It could likely build on top of `__cuda_array_interface__` objects.
